cluster/images/etcd/migrate-if-needed.sh

```diff
@@ -163,6 +114,90 @@ if [ "${CURRENT_VERSION}" = "2.2.1" -a "${CURRENT_VERSION}" != "${TARGET_VERSION
   echo "Backup done in ${BACKUP_DIR}"
 fi
 
+# Rollback to previous minor version of etcd 3.x, if needed.
+# This approach has only been tested with 3.2 and earlier. If newer versions of etcd support this
+# downgrade approach, please update the "${CURRENT_MINOR_VERSION} -le ..." check here.
+#
+# Warning: For HA etcd clusters (any cluster with more than one member), all members must be stopped before rolling back, zero
+# downtime rollbacks are not supported.
+rollback_etcd3_minor_version() {
+  if [ ${TARGET_MINOR_VERSION} != $((${CURRENT_MINOR_VERSION}-1)) ]; then
+    echo "Rollback from ${CURRENT_VERSION} to ${TARGET_VERSION} not supported, only rollbacks to the previous minor version are supported."
+    exit 1
+  fi
+  echo "Performing etcd ${CURRENT_VERSION} -> ${TARGET_VERSION} rollback"
+  ROLLBACK_BACKUP_DIR="${DATA_DIRECTORY}.bak"
+  rm -rf "${ROLLBACK_BACKUP_DIR}"
+  SNAPSHOT_FILE="${DATA_DIRECTORY}.snapshot.db"
+  rm -rf "${SNAPSHOT_FILE}"
+  ETCD_CMD="/usr/local/bin/etcd-${CURRENT_VERSION}"
+  ETCDCTL_CMD="/usr/local/bin/etcdctl-${CURRENT_VERSION}"
+
+  # Start CURRENT_VERSION of etcd.
+  START_VERSION="${CURRENT_VERSION}"
+  START_STORAGE="${CURRENT_STORAGE}"
+  echo "Starting etcd version ${START_VERSION} to capture rollback snapshot."
+  if ! start_etcd; then
+    echo "Unable to automatically downgrade etcd: starting etcd version ${START_VERSION} to capture rollback snapshot failed."
+    echo "See https://coreos.com/etcd/docs/3.2.13/op-guide/recovery.html for manual downgrade options."
+    exit 1
+  else
+    ETCDCTL_API=3 ${ETCDCTL_CMD} snapshot --endpoints "http://127.0.0.1:${ETCD_PORT}" save "${SNAPSHOT_FILE}"
+  fi
+  stop_etcd
+
+  # Backup the data before rolling back.
+  mv "${DATA_DIRECTORY}" "${ROLLBACK_BACKUP_DIR}"
+  ETCDCTL_CMD="/usr/local/bin/etcdctl-${TARGET_VERSION}"
+  NAME="etcd-$(hostname)"
+  ETCDCTL_API=3 ${ETCDCTL_CMD} snapshot restore "${SNAPSHOT_FILE}" \
+      --data-dir "${DATA_DIRECTORY}"  --name "${NAME}" --initial-cluster "${NAME}=http://localhost:2380"
+
+  CURRENT_VERSION="${TARGET_VERSION}"
+  echo "${CURRENT_VERSION}/${CURRENT_STORAGE}" > "${DATA_DIRECTORY}/${VERSION_FILE}"
+}
+
+# Rollback from "3.0.x" version in 'etcd3' mode to "2.2.1" version in 'etcd2' mode, if needed.
+rollback_to_etcd2() {
```

Comment:
Body of this function is unchanged. The code was moved from inline script (below) to this function.